README.rst
cwave - Continous Wavelet Transform (CWT) library
=================================================

cwave is a Python2/3 library implementing the Continuous Wavelet Transform (CWT)
using FFT as in `A Practical Guide to Wavelet Analysis` by C. Torrence and G.P.
Compo. cwave is an open-source, GPLv3-licensed software.

.. image:: https://travis-ci.org/compbio-fmach/cwave.png?branch=master
  :target: https://travis-ci.org/compbio-fmach/cwave

.. image:: https://readthedocs.org/projects/cwave/badge/?version=latest
  :target: http://cwave.readthedocs.io/en/latest/?badge=latest
  :alt: Documentation Status

* `Homepage and documentation <http://cwave.readthedocs.io>`_
* `Issues <https://github.com/compbio-fmach/cwave/issues>`_
* `Github page <https://github.com/compbio-fmach/cwave>`_

Financial Contributions
^^^^^^^^^^^^^^^^^^^^^^^
`Computational Biology Unit - Research and Innnovation Center at Fondazione
Edmund Mach <http://www.fmach.it/eng>`_

CHANGES.rst
CHANGES
=======

1.0
---
Initial release

cwave/trans.py
##    Copyright 2017 cwave Developers <davide.albanese@gmail.com>

##    This file is part of cwave.
##
##    cwave is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    cwave is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.

##    You should have received a copy of the GNU General Public License
##    along with cwave. If not, see <http://www.gnu.org/licenses/>.

from __future__ import division

import copy

import numpy as np
import scipy as sp

from . import wavefun


class Series(object):
    """Series.

    Example
    -------
    >>> import cwave
    >>> import numpy as np
    >>> x = np.random.sample(16)
    >>> dt=2
    >>> s = cwave.Series(x, dt)
    """

    def __init__(self, x, dt=1):
        """
        Parameters
        ----------
        x : 1d array_like object
            series (N).
        dt : float
            time step.
        """

        x_arr = np.array(x, dtype=np.float, copy=True)
        if x_arr.ndim != 1:
            raise ValueError('x must be an 1d array_like object')
        
        if dt <= 0.0:
            raise ValueError('dt must be > 0')
     
        self._x = x_arr
        self._dt = dt

    @property
    def x(self):
        """Get the series (1d numpy array).
        """
        
        return self._x
    
    @property
    def dt(self):
        """Get the time step (float).
        """

        return self._dt


class Trans(object):
    """Continuous wavelet transformed series.
    """

    def __init__(self, W, scales, omega, wavelet, x_mean=0, dt=1):
        """
        Parameters
        ----------
        W : 2d array_like object
            transformed series (J x N).
        scales : 1d array_like object
            wavelet scales (J).
        omega : 1d array_like object
            angular frequencies (N).
        wavelet : Wavelet object
            wavelet used for transformation.
        x_mean : float
            mean value of the original (non-transformed) series.
        dt : float
            time step.        
        """

        W_arr = np.array(W, copy=True)
        if W_arr.ndim != 2:
            raise ValueError('W must be an 2d array_like object')

        scales_arr = np.array(scales, dtype=np.float, copy=True)
        if scales_arr.ndim != 1:
            raise ValueError('scales must be an 1d array_like object')

        omega_arr = np.array(omega, dtype=np.float, copy=True)
        if omega_arr.ndim != 1:
            raise ValueError('omega must be an 1d array_like object')

        if (W_arr.shape[0] != scales_arr.shape[0]) or \
            (W_arr.shape[1] != omega_arr.shape[0]):
            raise ValueError('shape of W must be (n_scales, n_omega)')
        
        if not isinstance(wavelet, wavefun.Wavelet):
            raise ValueError('wavelet must be an Wavelet object')

        if dt <= 0.0:
            raise ValueError('dt must be > 0')

        self._W = W_arr
        self._scales = scales_arr
        self._omega = omega_arr
        self._wavelet = copy.deepcopy(wavelet)
        self._x_mean = x_mean
        self._dt = dt

    @property
    def W(self):
        """Get the transformed series (2d numpy array).
        """
        
        return self._W
    
    @property
    def scales(self):
        """Get the wavelet scales (1d numpy array).
        """

        return self._scales

    @property
    def omega(self):
        """Get the angular frequencies (1d numpy array).
        """

        return self._omega

    @property
    def wavelet(self):
        """Get the wavelet (Wavelet).
        """

        return self._wavelet

    @property
    def x_mean(self):
        """Get the mean value of the original (non-transformed) series (float).
        """

        return self._x_mean

    @property
    def dt(self):
        """Get the time step (float).
        """

        return self._dt

    def S(self):
        """Returns the wavelet power spectrum abs(W)^2."""

        return np.absolute(self._W)**2

    def var(self):
        """Returns the variance (eq. 14 in Torrence 1998)"""

        # (dj * dt) / Cd = psi0(0) * sqrt(dt) * Ck

        Ck = self._wavelet._factor_Ck(self._scales, self._omega, self._dt)
        psi00 = self._wavelet.time(0)

        return ((psi00 * np.sqrt(self._dt) * Ck) / self._W.shape[1]) * \
            np.sum(self.S() / self._scales.reshape(-1, 1)) 


def cwt(s, wavelet, dj=0.25, scale0=None, scales=None):
    """Continuous wavelet transform.
    
    Parameters
    ----------
    s : :class:`cwave.Series` object
        series.
    wavelet : :class:`cwave.Wavelet` object
        wavelet function.
    dj : float
        scale resolution (spacing between scales). A smaller dj will give better
        scale resolution. If scales is not None, this parameter is ignored.
    scale0 : float
        the smallest scale. If scale0=None it is chosen so that the 
        equivalent Fourier period is 2*dt. If scales is not None, this parameter
        is ignored.
    scales : None, float or 1d array_like object
        wavelet scale(s). If scales=None, the scales are automatically computed
        as fractional powers of two, with a scale resolution dj and the smallest
        scale scale0. If the parameter scales is provided, dj and scale0 are
        automatically ignored.

    Returns
    -------
    T : :class:`cwave.Trans` object
        continuous wavelet transformed series.

    Example
    -------
    >>> import cwave
    >>> import numpy as np
    >>> x = np.random.sample(8)
    >>> dt=2
    >>> T = cwave.cwt(cwave.Series(x, dt), cwave.DOG(), scales=[2, 4, 8])
    >>> T.S()
    array([[  1.19017195e-01,   1.54253543e-01,   9.07432163e-02,
              2.96227293e-02,   5.89519189e-04,   1.09486971e-01,
              1.15546678e-02,   3.93108223e-02],
           [  4.43627788e-01,   2.27266757e-01,   3.35649411e-05,
              1.86687786e-01,   3.78349904e-01,   2.24894861e-01,
              3.22011576e-03,   1.84538790e-01],
           [  8.01208492e-03,   4.40859411e-03,   1.92688516e-05,
              3.62275119e-03,   8.01206572e-03,   4.40859342e-03,
              1.92697929e-05,   3.62275056e-03]])
    """
    
    if not isinstance(wavelet, wavefun.Wavelet):
        raise ValueError('wavelet must be an Wavelet object')

    N = s.x.shape[0]
    dt = s.dt

    if scales is not None:
        scales_arr = np.atleast_1d(scales)
    else:
        scales_arr = wavelet.auto_scales(dt, dj, N, scale0)

    x_mean = np.mean(s.x)
    x_centered = s.x - x_mean
    
    omega = 2 * np.pi * np.fft.fftfreq(N, dt)
    x_hat = np.fft.fft(x_centered)

    W = []
    for scale in scales_arr:
        wavelet_hat = wavelet.freq(omega, scale, dt)
        W.append(np.fft.ifft(wavelet_hat * x_hat))
    
    return Trans(W, scales_arr, omega, wavelet, x_mean, dt)


def icwt(T):
    """Inverse continuous wavelet transform.

    Parameters
    ----------
    T : :class:`cwave.Trans` object
        continuous wavelet transformed series.

    Returns
    -------
    s : :class:`cwave.Series` object
        series.

    Example
    -------
    >>> import cwave
    >>> import numpy as np
    >>> x = np.random.normal(2, 1, 16)
    >>> x
    array([ 1.6960901 ,  3.27146653,  2.55896222,  2.39484518,  2.34977766,
            3.48575552,  1.7372688 , -0.21329766,  3.5425618 ,  3.34657898,
            1.54359934,  2.96181542,  2.43294205,  1.65980233,  3.44710306,
            1.69615204])
    >>> dt=1
    >>> T = cwave.cwt(cwave.Series(x, dt), cwave.DOG())
    >>> sr = cwave.icwt(T)
    >>> sr.x
    array([ 1.64067578,  3.28517018,  2.78434897,  2.38949828,  2.58014315,
            3.52751356,  1.34776275, -0.41078628,  3.3648406 ,  3.56461166,
            1.7286081 ,  2.88596331,  2.40275636,  1.81964648,  3.28932397,
            1.7113465 ])
    """

    if not isinstance(T, Trans):
        raise ValueError('W must be a cwave.Trans object')

    Ck = T.wavelet._factor_Ck(T.scales, T.omega, T.dt)
    x = Ck * np.sum(np.real(T.W) / np.sqrt(T.scales).reshape(-1, 1), axis=0)

    return Series(x + T.x_mean, T.dt)
    

examples/nino.py
# This code reproduces the figure 1a in (Torrence, 1998).
# This example does not include the zero padding at the end of the series and
# the red noise analysis.

from __future__ import division

import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec

import cwave

dt = 0.25
dj = 0.125

x = np.loadtxt("sst_nino3.txt")
s = cwave.Series(x, dt)

var = np.var(s.x)
N = s.x.shape[0]
times = dt * np.arange(N)

## Morlet
w_morlet = cwave.Morlet()
T_morlet = cwave.cwt(s, w_morlet, dj=dj, scale0=2*dt)
Sn_morlet= T_morlet.S() / var
# COI
taus_morlet= [T_morlet.wavelet.efolding_time(scale) for scale in T_morlet.scales]
mask_morlet = np.zeros_like(T_morlet.W, dtype=np.bool)
for i in range(mask_morlet.shape[0]):
    w = times < taus_morlet[i]
    mask_morlet[i, w] = True
    mask_morlet[i, w[::-1]] = True

## DOG
w_dog = cwave.DOG()
T_dog = cwave.cwt(s, w_dog, dj=dj, scale0=dt/2)
Sn_dog = T_dog.S() / var
# COI
taus_dog= [T_dog.wavelet.efolding_time(scale) for scale in T_dog.scales]
mask_dog = np.zeros_like(T_dog.W, dtype=np.bool)
for i in range(mask_dog.shape[0]):
    w = times < taus_dog[i]
    mask_dog[i, w] = True
    mask_dog[i, w[::-1]] = True

## Plot

fig = plt.figure(1)
gs = gridspec.GridSpec(3, 1, height_ratios=[0.6, 1, 1])

# plot the series s
ax1 = plt.subplot(gs[0])
p1 = ax1.plot(times, s.x, "k")

# plot the wavelet power spectrum (Morlet)
ax2 = plt.subplot(gs[1], sharex=ax1)
X, Y =  np.meshgrid(times, T_morlet.scales)
plt.contourf(X, Y, Sn_morlet, [1, 2, 5, 10], origin='upper', cmap=plt.cm.bone_r,
    extend='both')
plt.contour(X, Y, Sn_morlet, [1, 2, 5, 10], colors='k', linewidths=1, origin='upper')

# plot COI (Morlet)
plt.contour(X, Y, mask_morlet, [0, 1], colors='k', linewidths=1, origin='upper')
plt.contourf(X, Y, mask_morlet, [0, 1], colors='none', origin='upper', extend='both', 
    hatches=[None, 'x'])
ax2.set_yscale('log', basey=2)
plt.ylim(0.5, 64)
plt.gca().invert_yaxis()

# plot the wavelet power spectrum (DOG)
ax3 = plt.subplot(gs[2], sharex=ax1)
X, Y =  np.meshgrid(times, T_dog.scales)
plt.contourf(X, Y, Sn_dog, [2, 10], origin='upper', cmap=plt.cm.bone_r,
    extend='both')
plt.contour(X, Y, Sn_dog, [2, 10], colors='k', linewidths=1, origin='upper')

# plot COI (DOG)
plt.contour(X, Y, mask_dog, [0, 1], colors='k', linewidths=1, origin='upper')
plt.contourf(X, Y, mask_dog, [0, 1], colors='none', origin='upper', extend='both', 
    hatches=[None, 'x'])
ax3.set_yscale('log', basey=2)
plt.ylim(0.125, 16)
plt.gca().invert_yaxis()

plt.show()


examples/wavelet_fun.py
# This code reproduces the figure 2a in (Torrence, 1998).
# The plot on the left give the real part (solid) and the imaginary part
# (dashed) for the Morlet wavelet in the time domain. The plot on the right
# give the corresponding wavelet in the frequency domain.
# Change the wavelet function in order to obtain the figures 2b 2c and 2d.

from __future__ import division

import numpy as np
import scipy as sp
import matplotlib.pyplot as plt

import cwave


wavelet = cwave.Morlet() # Paul(), DOG() or DOG(m=6)
dt = 1
scale = 10*dt
t = np.arange(-40, 40, dt)
omega = np.arange(-1.2, 1.2, 0.01)

psi = wavelet.time(t, scale=scale, dt=dt)
psi_real = np.real(psi)
psi_imag = np.imag(psi)
psi_hat = wavelet.freq(omega, scale=scale, dt=dt)

fig = plt.figure(1)

ax1 = plt.subplot(121)
plt.plot(t/scale, psi_real, "k", label=("Real"))
plt.plot(t/scale, psi_imag, "k--", label=("Imag"))
plt.xlabel("t/s")
plt.legend()

ax2 = plt.subplot(122)
plt.plot((scale*omega)/(2*np.pi), psi_hat, "k")
plt.vlines(0, psi_hat.min(), psi_hat.max(), linestyles="dashed")
plt.xlabel("s * omega/(2 * PI)")

plt.show()


cwave/wavefun.py
##    Copyright 2017 cwave Developers <davide.albanese@gmail.com>

##    This file is part of cwave.
##
##    cwave is free software: you can redistribute it and/or modify
##    it under the terms of the GNU General Public License as published by
##    the Free Software Foundation, either version 3 of the License, or
##    (at your option) any later version.
##
##    cwave is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.

##    You should have received a copy of the GNU General Public License
##    along with cwave. If not, see <http://www.gnu.org/licenses/>.

from __future__ import division

import abc

import six

import numpy as np
import scipy as sp
import scipy.misc
import scipy.special


@six.add_metaclass(abc.ABCMeta)
class Wavelet(object):
    """The abstract wavelet base class.
    """
    
    @abc.abstractmethod
    def time(self, t, scale=1, dt=None):
        """Wavelet function in the time domain psi0(t/s).

        Parameters
        ----------
        t : float
            time. If df is not None each element of t should be multiple of dt.
        scale : float
            wavelet scale.
        dt : None or float
            time step. If dt is not None, the method returns the 
            energy-normalized psi (psi0). 

        Returns
        -------
        psi0 : float/complex or 1D numpy array
            the wavelet in the time domain.

        Example
        -------
        >>> import cwave
        >>> w = cwave.Morlet()
        >>> t = range(-8, 8)
        >>> w.time(t, 10)
        array([ 0.04772449+0.54333716j, -0.28822893+0.51240757j,
               -0.56261977+0.27763414j, -0.65623233-0.09354365j,
               -0.51129130-0.46835014j, -0.16314795-0.69929479j,
                0.26678672-0.68621589j,  0.61683875-0.42200209j,
                0.75112554+0.j        ,  0.61683875+0.42200209j,
                0.26678672+0.68621589j, -0.16314795+0.69929479j,
               -0.51129130+0.46835014j, -0.65623233+0.09354365j,
               -0.56261977-0.27763414j, -0.28822893-0.51240757j])
        """

    @abc.abstractmethod
    def freq(self, omega, scale=1, dt=None):
        """Wavelet function in the frequency domain psi0_hat(s*omega).

        Parameters
        ----------
        omega : float or 1d array_like object
            angular frequency (length N).
        scale : float
            wavelet scale.
        dt : None or float
            time step. If dt is not None, the method returns the 
            energy-normalized psi_hat (psi_hat0).

        Returns
        -------
        psi0_hat : float/complex or 1D numpy array
            the wavelet in the frequency domain

        Example
        -------
        >>> import numpy as np
        >>> import cwave
        >>> w = cwave.Morlet()
        >>> omega = 2 * np.pi * np.fft.fftfreq(32, 1)
        >>> w.freq(omega, 10)
        array([  0.00000000e+000,   2.17596717e-004,   8.76094852e-002,
                 7.46634798e-001,   1.34686366e-001,   5.14277294e-004,
                 4.15651521e-008,   7.11082035e-014,   2.57494732e-021,
                 1.97367345e-030,   3.20214178e-041,   1.09967442e-053,
                 7.99366604e-068,   1.22994640e-083,   4.00575772e-101,
                 2.76147731e-120,   0.00000000e+000,   0.00000000e+000,
                 0.00000000e+000,   0.00000000e+000,   0.00000000e+000,
                 0.00000000e+000,   0.00000000e+000,   0.00000000e+000,
                 0.00000000e+000,   0.00000000e+000,   0.00000000e+000,
                 0.00000000e+000,   0.00000000e+000,   0.00000000e+000,
                 0.00000000e+000,   0.00000000e+000])
        """

    @abc.abstractmethod
    def fourier_period(self, scale):
        """Computes the equivalent Fourier period (wavelength) from wavelet
        scale.

        Parameters
        ----------
        scale : float
            wavelet scale.

        Returns
        -------
        lmbd : float
            equivalent Fourier period.

        Example
        -------
        >>> import cwave
        >>> w = cwave.Morlet()
        >>> w.fourier_period(scale=1)
        1.0330436477492537
        """

    def wavelet_scale(self, lmbd):
        """Computes the wavelet scale from equivalent Fourier period
        (wavelength).

        Parameters
        ----------
        lmbd : float
            equivalent Fourier period.

        Returns
        -------
        scale : float
            wavelet scale.

        Example
        -------
        >>> import cwave
        >>> w = cwave.Morlet()
        >>> w.wavelet_scale(lmbd=10)
        9.6801330919439117
        """
      
        scale = self.fourier_period(1)**(-1) * lmbd

        return scale

    @abc.abstractmethod
    def efolding_time(self, scale):
        """Returns the e-folding time tau_s.

        Parameters
        ----------
        scale : float
            wavelet scale.

        Returns
        -------
        tau : float
            e-folding time.

        Example
        -------
        >>> import cwave
        >>> w = cwave.Morlet()
        >>> w.efolding_time(1)
        1.4142135623730951
        """

    def smallest_scale(self, dt):
        """Returns the smallest resolvable scale. It is chosen so that the
        equivalent Fourier period is 2*dt.

        Parameters
        ----------
        dt : float
            time step.

        Returns
        -------
        scale0 : float
            smallest scale.

        Example
        -------
        >>> import cwave
        >>> w = cwave.Morlet()
        >>> w.smallest_scale(dt=1)
        1.9360266183887822
        """

        scale0 = self.wavelet_scale(2*dt)

        return scale0


    def auto_scales(self, dt, dj, N, scale0=None):
        """Computes the equivalent Fourier period (wavelength) from wavelet
        scale.

        Parameters
        ----------
        dt : float
            time step.
        dj : float
            scale resolution. For the Morlet wavelet, a dj of about 0.5 is the
            largest value that still gives adequate sampling in scale, while for
            the other wavelet functions al larger value can be used.
        N : integer
            number of data samples.
        scale0 : float
            the smallest scale. If scale0=None it is chosen so that the 
            equivalent Fourier period is 2*dt.

        Returns
        -------
        scale : 1d numpy array
            scales.

        Example
        -------
        >>> import cwave
        >>> w = cwave.Morlet()
        >>> w.auto_scales(dt=1, dj=0.125, N=64, scale0=1)
        array([  1.        ,   1.09050773,   1.18920712,   1.29683955,
                 1.41421356,   1.54221083,   1.68179283,   1.83400809,
                 2.        ,   2.18101547,   2.37841423,   2.59367911,
                 2.82842712,   3.08442165,   3.36358566,   3.66801617,
                 4.        ,   4.36203093,   4.75682846,   5.18735822,
                 5.65685425,   6.1688433 ,   6.72717132,   7.33603235,
                 8.        ,   8.72406186,   9.51365692,  10.37471644,
                 11.3137085 ,  12.3376866 ,  13.45434264,  14.67206469,
                 16.        ,  17.44812372,  19.02731384,  20.74943287,
                 22.627417  ,  24.67537321,  26.90868529,  29.34412938,
                 32.        ,  34.89624745,  38.05462768,  41.49886575,
                 45.254834  ,  49.35074641,  53.81737058,  58.68825877,  64.        ])
        """

        if scale0 is None:
            scale0 = self.smallest_scale(dt)
            
        J = int((1 / dj) * np.log2((N * dt) / scale0))
        scale = scale0 * 2**(dj * np.arange(J+1))

        return scale

    def _factor_Ck(self, scales, omega, dt):
        """The reconstruction factor Ck is defined as 
        (dj*sqrt(dt))/(Cd*psi0(0)) = 1/(sum(real(Wd(sj))/sqrt(sj))
        (see eq. 13 in Torrence 1998).

        Parameters
        ----------
        omega : float or 1d array_like object
            angular frequency(es).
        scale : float or 1d array_like object
            wavelet scale(s).
        dt : float
            time step.

        Returns
        -------
        Ck : 1d numpy array
            reconstruction factor Ck.
        """

        scales_arr = np.atleast_1d(scales)
        if scales_arr.ndim > 1:
            raise ValueError("scales must be float or an 1d array_like object")

        omega_arr = np.atleast_1d(omega)
        if omega_arr.ndim > 1:
            raise ValueError("omega must be float or an 1d array_like object")

        if dt <= 0:
            raise ValueError('dt must be > 0')

        N = omega_arr.shape[0]
        
        d = 0.
        for scale in scales_arr:
            Wd = self.freq(omega_arr, scale, dt).sum() / N
            d += np.sum(np.real(Wd) / np.sqrt(scale))
        Ck = 1 / d

        return Ck

    def _factor_Cd(self, scales, omega, dt, dj):
        """The factor Cd is defined in eq. 13 in (Torrence 1998).

        Parameters
        ----------
        omega : float or 1d array_like object
            angular frequency(es).
        scale : float or 1d array_like object
            wavelet scale(s).
        dt : integer
            number of data samples.
        dj : float
            scale resolution.

        Returns
        -------
        Ck : 1d numpy array
            reconstruction factor Ck.
        """

        if dj <= 0:
            raise ValueError('dj must be > 0')

        Ck = self._factor_Ck(scales, omega, dt)
        Cd = (dj * np.sqrt(dt)) / (Ck * self.time(0))

        return Cd

    def _norm_freq(self, scale, dt):
        return np.sqrt((2 * np.pi * scale) / dt)

    def _norm_time(self, scale, dt):
        return np.sqrt(dt / scale)


class Morlet(Wavelet):
    """Morlet wavelet function.

    Example
    -------
    >>> import cwave
    >>> w = cwave.Morlet(omega0=6)
    """
    
    def __init__(self, omega0=6):
        """
        Parameters
        ----------
        omega0 : float
            nondimesional frequency. It should be >=6 in order to satisfy
            the admissibility condition (see Torrence 1998, Farge 1992).
        """
        
        self._omega0 = omega0

    @property
    def omega0(self):
        """Get the omega0 parameter"""

        return self._omega0

    def time(self, t, scale=1, dt=None):

        t_arr = np.asarray(t)
        if t_arr.ndim > 1:
            raise ValueError("t must be a float or an 1d array_like object")

        eta = t_arr / scale

        psi = np.pi**(-0.25) * np.exp(1.j * self._omega0 * eta) * \
            np.exp(-0.5 * (eta**2))

        if dt is not None:
            psi *= self._norm_time(scale, dt)

        return psi
    time.__doc__ = Wavelet.time.__doc__

    def freq(self, omega, scale=1, dt=None):

        omega_arr = np.asarray(omega)
        if omega_arr.ndim > 1:
            raise ValueError("omega must be a float or an 1d array_like object")

        psi_hat = np.zeros(omega_arr.shape[0], dtype=np.float)

        H = (omega_arr > 0)
        k = omega_arr[H] * scale
        psi_hat[H] = np.pi**(-0.25) * np.exp(-(k - self._omega0)**2 / 2.)

        if dt is not None:
            psi_hat *= self._norm_freq(scale, dt)

        return psi_hat
    freq.__doc__ = Wavelet.freq.__doc__

    def fourier_period(self, scale):

        lmbd = (4 * np.pi * scale) / \
            (self._omega0 + np.sqrt(2 + self._omega0**2))

        return lmbd
    fourier_period.__doc__ = Wavelet.fourier_period.__doc__

    def efolding_time(self, scale):

        tau = np.sqrt(2) * scale

        return tau
    efolding_time.__doc__ = Wavelet.efolding_time.__doc__


class Paul(Wavelet):
    """Paul Wavelet function.

    Example
    -------
    >>> import cwave
    >>> w = cwave.Paul(m=4)
    """

    def __init__(self, m=4):
        """
        Parameters
        ----------
        m : float
            order.
        """
        
        self._m = m

    @property
    def m(self):
        """Get the m parameter (order)"""

        return self._m

    def time(self, t, scale=1, dt=None):

        t_arr = np.asarray(t)
        if t_arr.ndim > 1:
            raise ValueError("t must be a float or an 1d array_like object")

        eta = t_arr / scale

        psi = (1- 1j*eta)**(-(self._m+1)) * \
            (2**self._m * 1j**self._m * scipy.misc.factorial(self._m)) / \
            np.sqrt(np.pi * scipy.misc.factorial(2*self._m))

        if dt is not None:
            psi *= self._norm_time(scale, dt)

        return psi
    time.__doc__ = Wavelet.time.__doc__

    def freq(self, omega, scale=1, dt=None):

        omega_arr = np.asarray(omega)
        if omega_arr.ndim > 1:
            raise ValueError("omega must be a float or an 1d array_like object")

        psi_hat = np.zeros(omega_arr.shape[0], dtype=np.float)

        H = (omega_arr > 0)
        k = omega_arr[H] * scale
        psi_hat[H] = k**self._m * np.exp(-k) * 2**self._m / \
            np.sqrt(self._m * scipy.misc.factorial(2 * self._m - 1))

        if dt is not None:
            psi_hat *= self._norm_freq(scale, dt)

        return psi_hat
    freq.__doc__ = Wavelet.freq.__doc__

    def fourier_period(self, scale):

        lmbd = (4 * np.pi * scale) / (2 * self._m + 1)

        return lmbd
    fourier_period.__doc__ = Wavelet.fourier_period.__doc__

    def efolding_time(self, scale):

        tau = scale / np.sqrt(2)

        return tau
    efolding_time.__doc__ = Wavelet.efolding_time.__doc__


class DOG(Wavelet):
    """Derivative Of Gaussian (DOG) Wavelet function.

    Example
    -------
    >>> import cwave
    >>> w = cwave.DOG(m=2)
    """

    def __init__(self, m=2):
        """
        Parameters
        ----------
        m : float
            derivative.
        """
        
        self._m = m

    @property
    def m(self):
        """Get the m parameter (derivative)"""

        return self._m

    def time(self, t, scale=1, dt=None):

        t_arr = np.asarray(t)
        if t_arr.ndim > 1:
            raise ValueError("t must be a float or an 1d array_like object")

        eta = t_arr / scale

        hn = scipy.special.hermitenorm(self._m)
        psi = -(hn(eta) * np.exp(-eta**2/2)) / \
            np.sqrt(scipy.special.gamma(self._m+.5))

        if dt is not None:
            psi *= self._norm_time(scale, dt)

        return psi
    time.__doc__ = Wavelet.time.__doc__

    def freq(self, omega, scale=1, dt=None):

        omega_arr = np.asarray(omega)
        if omega_arr.ndim > 1:
            raise ValueError("omega must be a float or an 1d array_like object")

        k = omega_arr * scale
        
        psi_hat = k**self._m * np.exp(-k**2 / 2) / \
            np.sqrt(scipy.special.gamma(self._m+.5))
        
        if dt is not None:
            psi_hat *= self._norm_freq(scale, dt)

        return psi_hat
    freq.__doc__ = Wavelet.freq.__doc__

    def fourier_period(self, scale):

        lmbd = (2 * np.pi * scale) / np.sqrt(self._m + .5)

        return lmbd
    fourier_period.__doc__ = Wavelet.fourier_period.__doc__

    def efolding_time(self, scale):

        tau = np.sqrt(2) * scale

        return tau
    efolding_time.__doc__ = Wavelet.efolding_time.__doc__
